File: integrations/ue/Plugins/VideoReview/Source/VideoReviewEditor/Private/VideoReviewAPI.cpp
#include "VideoReviewAPI.h"
#include "VideoReviewAPIExecute.h"
#include "Dom/JsonObject.h"
#include "Dom/JsonValue.h"
#include "Serialization/JsonReader.h"
#include "Serialization/JsonSerializer.h"

FString FVideoReviewAPI::ServerUrl;
FString FVideoReviewAPI::ApiToken;
int32 FVideoReviewAPI::TimeoutMs = 300000;
FString FVideoReviewAPI::CliRootOverride;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

TVideoReviewResponse<FVideoReviewGenericResult> FVideoReviewAPI::Bootstrap(const FString& Email, const FString& Password)
{
    return ExecuteObject(TEXT("bootstrap"), { TEXT("--email"), Email, TEXT("--pass"), Password });
}

TVideoReviewResponse<FVideoReviewGenericResult> FVideoReviewAPI::CreateUser(const FString& Name, const FString& Email, const FString& Password)
{
    return ExecuteObject(TEXT("create-user"), { TEXT("--name"), Name, TEXT("--email"), Email, TEXT("--pass"), Password });
}

TVideoReviewResponse<TArray<FVideoReviewVideo>> FVideoReviewAPI::GetVideos(const FString& FilterTree, bool bIncludeRevisions)
{
    TArray<FString> Args = {
        TEXT("--include_revisions"), bIncludeRevisions ? TEXT("true") : TEXT("false"),
        TEXT("--filter_tree"), FilterTree,
    };
    return ExecuteVideoArray(TEXT("get-videos"), Args);
}

TVideoReviewResponse<TArray<FVideoReviewVideoRevision>> FVideoReviewAPI::GetVideosRev(const FString& VideoId)
{
    return ExecuteRevisionArray(TEXT("get-videos-rev"), { TEXT("--video_id"), VideoId });
}

TVideoReviewResponse<TArray<FVideoReviewComment>> FVideoReviewAPI::GetComments(const FString& VideoId)
{
    return ExecuteCommentArray(TEXT("get-comments"), { TEXT("--video_id"), VideoId });
}

TVideoReviewResponse<FVideoReviewVideoRevision> FVideoReviewAPI::UploadVideo(
    const FString& Title,
    const FString& FolderKey,
    const FString& VideoPath,
    const FString& ScenePath,
    const TArray<FString>& VcsWatchPaths)
{
    TArray<FString> Args = {
        TEXT("--title"), Title,
        TEXT("--folder_key"), FolderKey,
        TEXT("--video_path"), VideoPath,
    };
    if (!ScenePath.IsEmpty())
    {
        Args.Add(TEXT("--scene_path"));
        Args.Add(ScenePath);
    }
    if (VcsWatchPaths.Num() > 0)
    {
        Args.Add(TEXT("--vcs_watch_paths"));
        Args.Add(FString::Join(VcsWatchPaths, TEXT(",")));
    }
    return ExecuteRevisionObject(TEXT("upload-video"), Args);
}

TVideoReviewResponse<FVideoReviewSimpleMessage> FVideoReviewAPI::CreateVideoThumbnail(const FString& VideoId)
{
    return ExecuteText(TEXT("create-video-tmb"), { TEXT("--video_id"), VideoId });
}

TVideoReviewResponse<FVideoReviewGenericResult> FVideoReviewAPI::DeleteVideo(const FString& VideoId)
{
    return ExecuteObject(TEXT("delete-video"), { TEXT("--video_id"), VideoId });
}

TVideoReviewResponse<FVideoReviewGenericResult> FVideoReviewAPI::PurgeRev(const FString& VideoId, int32 Revision)
{
    return ExecuteObject(TEXT("purge-revision"), { TEXT("--video_id"), VideoId, TEXT("--revision"), FString::FromInt(Revision) });
}

TVideoReviewResponse<FVideoReviewAnnotateResult> FVideoReviewAPI::AnnotateVideoRev(const FString& VideoRevId)
{
    FString Id = VideoRevId.IsEmpty() ? TEXT("all") : VideoRevId;
    return ExecuteAnnotateObject(TEXT("annotate-video-rev"), { TEXT("--video_rev_id"), Id });
}

TVideoReviewResponse<FVideoReviewSimpleMessage> FVideoReviewAPI::UploadEventContext(const FString& VideoRevId, const FString& JsonPath)
{
    return ExecuteText(TEXT("upload-video-event-context"), { TEXT("--video_rev_id"), VideoRevId, TEXT("--json_path"), JsonPath });
}

TVideoReviewResponse<FVideoReviewSimpleMessage> FVideoReviewAPI::PatchVideo(const FString& VideoId, const TArray<FString>& VcsWatchPaths)
{
    TArray<FString> Args = { TEXT("--video_id"), VideoId };
    if (VcsWatchPaths.Num() > 0)
    {
        Args.Add(TEXT("--vcs_watch_paths"));
        Args.Add(FString::Join(VcsWatchPaths, TEXT(",")));
    }
    return ExecuteText(TEXT("patch-video"), Args);
}

// ---------------------------------------------------------------------------
// Internal execution helpers
// ---------------------------------------------------------------------------

FVideoReviewCommandResult FVideoReviewAPI::Execute(const FString& Command, const TArray<FString>& Args)
{
    return FVideoReviewAPIExecute::Run(Command, Args, ServerUrl, ApiToken, TimeoutMs, CliRootOverride);
}

TVideoReviewResponse<FVideoReviewSimpleMessage> FVideoReviewAPI::ExecuteText(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<FVideoReviewSimpleMessage> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;
    Response.Data.Text = Response.CommandResult.StdOut;
    return Response;
}

TVideoReviewResponse<FVideoReviewGenericResult> FVideoReviewAPI::ExecuteObject(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<FVideoReviewGenericResult> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;

    if (!Response.bSuccess) return Response;

    TSharedPtr<FJsonObject> JsonObject;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response.CommandResult.StdOut);
    if (!FJsonSerializer::Deserialize(Reader, JsonObject) || !JsonObject.IsValid())
    {
        Response.bSuccess = false;
        Response.ParseError = TEXT("Failed to parse JSON object");
        return Response;
    }

    JsonObject->TryGetBoolField(TEXT("success"), Response.Data.bSuccess);
    JsonObject->TryGetBoolField(TEXT("ok"), Response.Data.bOk);
    JsonObject->TryGetStringField(TEXT("error"), Response.Data.Error);
    JsonObject->TryGetStringField(TEXT("warning"), Response.Data.Warning);
    JsonObject->TryGetStringField(TEXT("token"), Response.Data.Token);
    JsonObject->TryGetStringField(TEXT("videoId"), Response.Data.VideoId);
    JsonObject->TryGetNumberField(TEXT("revision"), Response.Data.Revision);
    return Response;
}

TVideoReviewResponse<FVideoReviewAnnotateResult> FVideoReviewAPI::ExecuteAnnotateObject(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<FVideoReviewAnnotateResult> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;

    if (!Response.bSuccess) return Response;

    TSharedPtr<FJsonObject> JsonObject;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response.CommandResult.StdOut);
    if (!FJsonSerializer::Deserialize(Reader, JsonObject) || !JsonObject.IsValid())
    {
        Response.bSuccess = false;
        Response.ParseError = TEXT("Failed to parse JSON object");
        return Response;
    }

    JsonObject->TryGetNumberField(TEXT("successCount"), Response.Data.SuccessCount);
    JsonObject->TryGetNumberField(TEXT("failureCount"), Response.Data.FailureCount);
    return Response;
}

TVideoReviewResponse<FVideoReviewVideoRevision> FVideoReviewAPI::ExecuteRevisionObject(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<FVideoReviewVideoRevision> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;

    if (!Response.bSuccess) return Response;

    TSharedPtr<FJsonObject> JsonObject;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response.CommandResult.StdOut);
    if (!FJsonSerializer::Deserialize(Reader, JsonObject) || !JsonObject.IsValid())
    {
        Response.bSuccess = false;
        Response.ParseError = TEXT("Failed to parse JSON object");
        return Response;
    }

    Response.Data = ParseRevision(JsonObject);
    return Response;
}

TVideoReviewResponse<TArray<FVideoReviewVideo>> FVideoReviewAPI::ExecuteVideoArray(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<TArray<FVideoReviewVideo>> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;

    if (!Response.bSuccess) return Response;

    TArray<TSharedPtr<FJsonValue>> JsonArray;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response.CommandResult.StdOut);
    if (!FJsonSerializer::Deserialize(Reader, JsonArray))
    {
        Response.bSuccess = false;
        Response.ParseError = TEXT("Failed to parse JSON array");
        return Response;
    }

    for (const TSharedPtr<FJsonValue>& Value : JsonArray)
    {
        const TSharedPtr<FJsonObject>* Obj;
        if (Value->TryGetObject(Obj))
        {
            Response.Data.Add(ParseVideo(*Obj));
        }
    }
    return Response;
}

TVideoReviewResponse<TArray<FVideoReviewVideoRevision>> FVideoReviewAPI::ExecuteRevisionArray(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<TArray<FVideoReviewVideoRevision>> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;

    if (!Response.bSuccess) return Response;

    TArray<TSharedPtr<FJsonValue>> JsonArray;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response.CommandResult.StdOut);
    if (!FJsonSerializer::Deserialize(Reader, JsonArray))
    {
        Response.bSuccess = false;
        Response.ParseError = TEXT("Failed to parse JSON array");
        return Response;
    }

    for (const TSharedPtr<FJsonValue>& Value : JsonArray)
    {
        const TSharedPtr<FJsonObject>* Obj;
        if (Value->TryGetObject(Obj))
        {
            Response.Data.Add(ParseRevision(*Obj));
        }
    }
    return Response;
}

TVideoReviewResponse<TArray<FVideoReviewComment>> FVideoReviewAPI::ExecuteCommentArray(const FString& Command, const TArray<FString>& Args)
{
    TVideoReviewResponse<TArray<FVideoReviewComment>> Response;
    Response.CommandResult = Execute(Command, Args);
    Response.bSuccess = Response.CommandResult.bSuccess;

    if (!Response.bSuccess) return Response;

    TArray<TSharedPtr<FJsonValue>> JsonArray;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response.CommandResult.StdOut);
    if (!FJsonSerializer::Deserialize(Reader, JsonArray))
    {
        Response.bSuccess = false;
        Response.ParseError = TEXT("Failed to parse JSON array");
        return Response;
    }

    for (const TSharedPtr<FJsonValue>& Value : JsonArray)
    {
        const TSharedPtr<FJsonObject>* Obj;
        if (Value->TryGetObject(Obj))
        {
            Response.Data.Add(ParseComment(*Obj));
        }
    }
    return Response;
}

// ---------------------------------------------------------------------------
// JSON parsing helpers
// ---------------------------------------------------------------------------

FVideoReviewVideoRevision FVideoReviewAPI::ParseRevision(const TSharedPtr<FJsonObject>& Obj)
{
    FVideoReviewVideoRevision Rev;
    Obj->TryGetStringField(TEXT("id"), Rev.Id);
    Obj->TryGetStringField(TEXT("videoId"), Rev.VideoId);
    Obj->TryGetNumberField(TEXT("revision"), Rev.Revision);
    Obj->TryGetStringField(TEXT("filePath"), Rev.FilePath);
    Obj->TryGetStringField(TEXT("uploadedAt"), Rev.UploadedAt);
    Obj->TryGetBoolField(TEXT("deleted"), Rev.bDeleted);
    Obj->TryGetStringField(TEXT("summary"), Rev.Summary);

    const TArray<TSharedPtr<FJsonValue>>* TagsArray;
    if (Obj->TryGetArrayField(TEXT("tags"), TagsArray))
    {
        for (const TSharedPtr<FJsonValue>& TagVal : *TagsArray)
        {
            FString Tag;
            if (TagVal->TryGetString(Tag)) Rev.Tags.Add(Tag);
        }
    }
    return Rev;
}

FVideoReviewVideo FVideoReviewAPI::ParseVideo(const TSharedPtr<FJsonObject>& Obj)
{
    FVideoReviewVideo Video;
    Obj->TryGetStringField(TEXT("id"), Video.Id);
    Obj->TryGetStringField(TEXT("title"), Video.Title);
    Obj->TryGetStringField(TEXT("folderKey"), Video.FolderKey);
    Obj->TryGetStringField(TEXT("scenePath"), Video.ScenePath);
    Obj->TryGetNumberField(TEXT("latestRevisionNum"), Video.LatestRevisionNum);
    Obj->TryGetBoolField(TEXT("deleted"), Video.bDeleted);

    const TArray<TSharedPtr<FJsonValue>>* RevsArray;
    if (Obj->TryGetArrayField(TEXT("revisions"), RevsArray))
    {
        for (const TSharedPtr<FJsonValue>& RevVal : *RevsArray)
        {
            const TSharedPtr<FJsonObject>* RevObj;
            if (RevVal->TryGetObject(RevObj)) Video.Revisions.Add(ParseRevision(*RevObj));
        }
    }
    return Video;
}

FVideoReviewComment FVideoReviewAPI::ParseComment(const TSharedPtr<FJsonObject>& Obj)
{
    FVideoReviewComment Comment;
    Obj->TryGetStringField(TEXT("id"), Comment.Id);
    Obj->TryGetStringField(TEXT("videoId"), Comment.VideoId);
    Obj->TryGetNumberField(TEXT("videoRevNum"), Comment.VideoRevNum);
    Obj->TryGetStringField(TEXT("userName"), Comment.UserName);
    Obj->TryGetStringField(TEXT("userEmail"), Comment.UserEmail);
    Obj->TryGetStringField(TEXT("comment"), Comment.Comment);
    Obj->TryGetNumberField(TEXT("time"), Comment.Time);
    Obj->TryGetStringField(TEXT("issueId"), Comment.IssueId);
    Obj->TryGetStringField(TEXT("drawingPath"), Comment.DrawingPath);
    Obj->TryGetStringField(TEXT("createdAt"), Comment.CreatedAt);
    Obj->TryGetStringField(TEXT("updatedAt"), Comment.UpdatedAt);
    Obj->TryGetBoolField(TEXT("deleted"), Comment.bDeleted);
    Obj->TryGetNumberField(TEXT("thumbsUp"), Comment.ThumbsUp);
    return Comment;
}

File: integrations/unity/VideoReviewUnity/Assets/VideoReview/Editor/API/VideoReviewAPI.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace VideoReview.Editor.API
{
    public static class VideoReviewAPI
    {
        public static string CliRootOverride = null;
        public static string ServerUrl { get; set; } = Environment.GetEnvironmentVariable("VIDEO_REVIEW_SERVER_URL") ?? string.Empty;
        public static string ApiToken { get; set; } =
            Environment.GetEnvironmentVariable("VIDEO_REVIEW_API_TOKEN")
            ?? Environment.GetEnvironmentVariable("ADMIN_MAINTENANCE_TOKEN")
            ?? string.Empty;

        public static int TimeoutMs { get; set; } = 300000;

        // Backward-compatible typo alias.
        public static VideoReviewResponse<VideoReviewGenericResult> Boostrap(string email, string pass)
        {
            return Bootstrap(email, pass);
        }

        public static VideoReviewResponse<VideoReviewGenericResult> Bootstrap(string email, string pass)
        {
            return ExecuteObject<VideoReviewGenericResult>(
                "bootstrap",
                "--email", email,
                "--pass", pass
            );
        }

        public static VideoReviewResponse<VideoReviewGenericResult> CreateUser(string name, string email, string pass)
        {
            return ExecuteObject<VideoReviewGenericResult>(
                "create-user",
                "--name", name,
                "--email", email,
                "--pass", pass
            );
        }

        public static VideoReviewResponse<VideoReviewVideo[]> GetVideos(string filterTree = null, bool includeRevisions = false)
        {
            var args = new List<string>
            {
                "--include_revisions",  includeRevisions ? "true" : "false",
                "--filter_tree", filterTree,
            };

            return ExecuteArray<VideoReviewVideo>("get-videos", args.ToArray());
        }

        public static VideoReviewResponse<VideoReviewVideoRevision[]> GetVideosRev(string videoID)
        {
            return ExecuteArray<VideoReviewVideoRevision>(
                "get-videos-rev",
                "--video_id", videoID
            );
        }

        public static VideoReviewResponse<VideoReviewGenericResult> DeleteVideo(string videoID)
        {
            return ExecuteObject<VideoReviewGenericResult>(
                "delete-video",
                "--video_id", videoID
            );
        }

        public static VideoReviewResponse<VideoReviewGenericResult> PurgeRev(string videoID, int revision = -1)
        {
            return ExecuteObject<VideoReviewGenericResult>(
                "purge-revision",
                "--video_id", videoID,
                "--revision", revision.ToString()
            );
        }

        public static VideoReviewResponse<VideoReviewVideoRevision> UploadVideo(string title, string folderKey, string scenePath, string videoPath, string[] vcsWatchPaths = null)
        {
            var args = new List<string>
            {
                "--title", title,
                "--folder_key", folderKey,
                "--video_path", videoPath,
            };

            if (!string.IsNullOrWhiteSpace(scenePath))
            {
                args.Add("--scene_path");
                args.Add(scenePath);
            }

            if (vcsWatchPaths != null && vcsWatchPaths.Length > 0)
            {
                args.Add("--vcs_watch_paths");
                args.Add(string.Join(",", vcsWatchPaths));
            }

            return ExecuteObject<VideoReviewVideoRevision>("upload-video", args.ToArray());
        }

        public static VideoReviewResponse<VideoReviewSimpleMessage> CreateVideoThumbnail(string videoID)
        {
            return ExecuteText(
                "create-video-tmb",
                "--video_id", videoID
            );
        }

        public static VideoReviewResponse<VideoReviewComment[]> GetComments(string videoID)
        {
            return ExecuteArray<VideoReviewComment>(
                "get-comments",
                "--video_id", videoID
            );
        }

        public static VideoReviewResponse<VideoReviewAnnotateResult> AnnotateVideoRev(string videoRevID)
        {
            return ExecuteObject<VideoReviewAnnotateResult>(
                "annotate-video-rev",
                "--video_rev_id", string.IsNullOrWhiteSpace(videoRevID) ? "all" : videoRevID
            );
        }

        public static VideoReviewResponse<VideoReviewSimpleMessage> UploadEventContext(string videoRevID, string jsonPath)
        {
            return ExecuteText(
                "upload-video-event-context",
                "--video_rev_id", videoRevID,
                "--json_path", jsonPath
            );
        }

        public static VideoReviewResponse<VideoReviewSimpleMessage> PatchVideo(string videoID, string[] vcsWatchPaths = null)
        {
            var args = new List<string>
            {
                "--video_id", videoID,
            };

            if (vcsWatchPaths != null && vcsWatchPaths.Length > 0)
            {
                args.Add("--vcs_watch_paths");
                args.Add(string.Join(",", vcsWatchPaths));
            }

            return ExecuteText("patch-video", args.ToArray());
        }

        private static VideoReviewResponse<T> ExecuteObject<T>(string command, params string[] args)
        {
            var result = Execute(command, args);
            var response = new VideoReviewResponse<T>
            {
                success = result.success,
                command = result,
                data = default(T),
                parseError = null,
            };

            if (!result.success)
            {
                return response;
            }

            if (VideoReviewJson.TryParseObject(result.stdOut, out T value, out var parseError))
            {
                response.data = value;
                return response;
            }

            response.parseError = parseError;
            response.success = false;
            return response;
        }

        private static VideoReviewResponse<T[]> ExecuteArray<T>(string command, params string[] args)
        {
            var result = Execute(command, args);
            var response = new VideoReviewResponse<T[]>
            {
                success = result.success,
                command = result,
                data = Array.Empty<T>(),
                parseError = null,
            };

            if (!result.success)
            {
                return response;
            }

            if (VideoReviewJson.TryParseArray(result.stdOut, out T[] values, out var parseError))
            {
                response.data = values;
                return response;
            }

            response.parseError = parseError;
            response.success = false;
            return response;
        }

        private static VideoReviewResponse<VideoReviewSimpleMessage> ExecuteText(string command, params string[] args)
        {
            var result = Execute(command, args);
            return new VideoReviewResponse<VideoReviewSimpleMessage>
            {
                success = result.success,
                command = result,
                data = new VideoReviewSimpleMessage { text = result.stdOut?.Trim() ?? string.Empty },
                parseError = null,
            };
        }

        private static VideoReviewCommandResult Execute(string command, params string[] args)
        {
            try
            {
                var result = VideoReviewAPIExecute.Run(
                    command,
                    args,
                    ServerUrl,
                    ApiToken,
                    TimeoutMs,
                    CliRootOverride
                );

                if (!result.success)
                {
                    Debug.LogError($"[VideoReview] CLI failed: {command}\n{result.stdErr}");
                }

                return result;
            }
            catch (Exception e)
            {
                Debug.LogError($"[VideoReview] CLI execution error: {e.Message}");
                return new VideoReviewCommandResult
                {
                    success = false,
                    timedOut = false,
                    exitCode = -1,
                    command = command,
                    stdOut = string.Empty,
                    stdErr = e.ToString(),
                };
            }
        }
    }
}
